File: NowCoder/Top101/BM100 设计LRU缓存结构.cpp
#include <map>

class Solution {
  public:
    Solution(int capacity) {
        this->capacity = capacity;
        this->count = 0;
        this->head = new node();
        this->tail = new node();
        this->head->next = this->tail;
        this->tail->prev = this->head;
    }

    int get(int key) {
        if (data.count(key)) {
            node* record = data[key];
            use(record);
            return record->value;
        }
        return -1;
    }

    void set(int key, int value) {
        if (data.count(key)) {
            node* record = data[key];
            record->value = value;
            use(record);
        } else {
            insert(key, value);
        }
    }

  private:
    struct node {
        node* prev = nullptr;
        node* next = nullptr;
        int key = 0;
        int value = 0;
        node(): {}
        node(node* prev, node* next, int key, int value):
            prev(prev), next(next), key(key), value(value) {}
    };

    map<int, node*>data;
    int capacity;
    int count;
    node* head;
    node* tail;

    void use(node* record) {
        record->prev->next = record->next;
        record->next->prev = record->prev;
        record->next = this->head->next;
        record->prev = this->head;
        record->next->prev = record;
        this->head->next = record;
    }

    void insert(int key, int value) {
        node* record = new node(this->head, this->head->next, key, value);
        this->data[key] = record;
        this->head->next->prev = record;
        this->head->next = record;
        this->count++;
        while (this->count > this->capacity) {
            erase();
        }
    }

    void erase() {
        node* least_used = this->tail->prev;
        this->tail->prev = least_used->prev;
        least_used->prev->next = this->tail;
        this->data.erase(least_used->key);
        delete least_used;
        this->count--;
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution* solution = new Solution(capacity);
 * int output = solution->get(key);
 * solution->set(key,value);
 */

File: NowCoder/Top101/BM100 设计LRU缓存结构.java
import java.util.HashMap;
import java.util.Map;

public class Solution {
    private final LRUCache<Integer, Integer> cache;

    public Solution(int capacity) {
        this.cache = new LRUCache<>(capacity);
    }

    public int get(int key) {
        Integer value = this.cache.get(key);
        return value == null ? -1 : value;
    }

    public void set(int key, int value) {
        this.cache.set(key, value);
    }

}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution solution = new Solution(capacity);
 * int output = solution.get(key);
 * solution.set(key,value);
 */

class CacheNode<K, V> {

    CacheNode<K, V> prev;
    CacheNode<K, V> next;
    K key;
    V value;

    public CacheNode() {

    }

    public CacheNode(CacheNode<K, V> prev, CacheNode<K, V> next, K key, V value) {
        this.prev = prev;
        this.next = next;
        this.key = key;
        this.value = value;
    }
}

abstract class BaseCache<K, V> {
    protected final Map<K, CacheNode<K, V>> data;
    protected final CacheNode<K, V> head;
    protected final CacheNode<K, V> tail;
    protected final int capacity;
    protected int count;

    protected BaseCache(int capacity) {
        this.data = new HashMap<>((int)(capacity * 1.0 / 0.75) + 1);
        this.head = new CacheNode<>();
        this.tail = new CacheNode<>();
        this.head.next = this.tail;
        this.tail.prev = this.head;
        this.capacity = capacity;
        this.count = 0;
    }

    public V get(K key) {
        if (data.containsKey(key)) {
            CacheNode<K, V> node = data.get(key);
            use(node);
            return node.value;
        }
        return null;
    }

    public void set(K key, V value) {
        if (data.containsKey(key)) {
            CacheNode<K, V> node = data.get(key);
            node.value = value;
            use(node);
        } else {
            insert(key, value);
        }
    }

    abstract void use(CacheNode<K, V> node);

    abstract void insert(K key, V value);

    abstract void erase();
}

class LRUCache<K, V> extends BaseCache<K, V> {
    public LRUCache(int capacity) {
        super(capacity);
    }

    @Override
    void use(CacheNode<K, V> node) {
        node.prev.next = node.next;
        node.next.prev = node.prev;

        node.next = head.next;
        node.prev = head;

        head.next.prev = node;
        head.next = node;
    }

    @Override
    void insert(K key, V value) {
        CacheNode<K, V> record = new CacheNode<>(head, head.next, key, value);
        data.put(key, record);
        head.next.prev = record;
        head.next = record;
        this.count++;
        while (this.count > this.capacity) {
            erase();
        }
    }

    @Override
    void erase() {
        CacheNode<K, V> leastUsed = tail.prev;
        tail.prev = leastUsed.prev;
        leastUsed.prev.next = tail;
        data.remove(leastUsed.key);
        this.count--;
    }
}
